src/schedule_synthesis.jl

```diff
@@ -0,0 +1,273 @@
+    K::AbstractMatrix{Float64}, z_0::AbstractVecOrMat, d_max::Float64,
+    maxwindow::Int64, n::Int64, t::Int64)::Vector{WeaklyHardConstraint}
+
+    # TODO: 
+    #   1) Test to make sure it works the same
+    #   2) Discuss: should additional safe constraints be included in result? 
+    #      Or only the ones that are actually checked & useful for scheduling
+    safe_constraints = MeetAny[]
+
+    # Do not need to go through all O(maxwindow^2) constraints,
+    # see paper for optimization argument
+    meet = 1
+    for window in 2:maxwindow
+        while meet < window
+            constraint = MeetAny(meet, window)
+            a = hold_kill(sysd, K, constraint)
+            # Check if the deviation bound is within the safety margin
+            reachable = bounded_runs_iter(a, z_0, n, t)
+            m = maximum(deviation(a, z_0, reachable))
+            if m <= d_max
+                # All constraints with (m, window) where m >= meet are valid
+                push!(safe_constraints, constraint)
+                break
+            end
+            meet += 1
+        end
+    end
+
+    safe_constraints
+end
+
+# Helper functions
+
+"""
+    _undigit(d[, base=2])
+
+Convert a list of digits to a number. Default base=2.
+```julia
+julia> _undigit([1, 0, 0])
+4
+```
+"""
+function _undigit(d::Vector{Int64}; base=2)::Int64
+    s = 0
+    mult = 1
+    for val in reverse(d)
+        s += val * mult
+        mult *= base
+    end
+    return s
+end
+
+"""
+    _digits_b2r(x[, pad])
+
+Digits **b**ase **2**, **r**everse
+A shortcut for `digits(x, base=2, pad=pad) |> reverse`
+"""
+function _digits_b2r(x::Int64, pad::Int64=0)::Vector{Int64}
+    digits(x, base=2, pad=pad) |> reverse
+end
+
+"""
+    _state_separation(l, B[, indigits=false])
+
+_state_separation takes a number `l` representing the overall state of
+a `_SynthesizedAutomaton` and an array `B` representing the number of bits
+for each comprising `Automaton` of the `_SynthesizedAutomaton` to compute a 
+list of individual states. For example
+
+```julia
+julia> _state_separation(6, [1, 2])
+[1, 2]
+```
+The explanation of the example is as follows
+```
+6 -> [1, 1, 0]
+[1, 1, 0] / [1, 2] -> [[1], [1, 0]]
+[[1], [1, 0]] -> [1, 2]
+```
+"""
+function _state_separation(l::Int64, B::Vector{Int64}; indigits=false)
+    @assert l < 2^sum(B) "l has more bits than the sum of B"
+    
+    bits = digits(l, base=2, pad=sum(B)) |> reverse
+    index_points = cumsum(B)
+    from_index = [0; index_points[1:end-1]]
+    to_index = index_points
+    bits_separated = [bits[i+1:j] for (i, j) in zip(from_index, to_index)]
+    if indigits
+        bits_separated
+    else
+        map(_undigit, bits_separated)
+    end
+end
+
+"""
+Struct for a controller automaton which enforces only the weakly 
+hard constraint for that controller. i.e., it is not concerned with the
+dynamical matrix of the system.
+"""
+struct _ConstraintAutomaton
```

Comment:
Just for my understanding, why is a separate type from the exported `Automaton` needed for this?  Seems like it creates some code duplication in the form of the `_controller_automaton` function.